staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -265,18 +460,196 @@ func (q *delayingType) waitingLoop() {
 }
 
 // insert adds the entry to the priority queue, or updates the readyAt if it already exists in the queue
-func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
+// It returns a boolean to indicate that the entry was inserted or not.
+func insert(q *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) bool {
 	// if the entry already exists, update the time only if it would cause the item to be queued sooner
 	existing, exists := knownEntries[entry.data]
 	if exists {
-		if existing.readyAt.After(entry.readyAt) {
-			existing.readyAt = entry.readyAt
-			heap.Fix(q, existing.index)
+		switch entry.options.WhenWaiting {
+		case TakeShorter:
+			if existing.readyAt.After(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
+		case TakeLonger:
+			if existing.readyAt.Before(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
+		case TakeIncoming:
+			if !existing.readyAt.Equal(entry.readyAt) {
+				existing.readyAt = entry.readyAt
+				existing.options = entry.options
+				heap.Fix(q, existing.index)
+				return true
+			}
+		}
+		return false
+	}
+
+	heap.Push(q, entry)
+	knownEntries[entry.data] = entry
+	return true
+}
+
+// calculateNextReadyAt calculates when the first item in the queue will be ready and
+// resets the nextReadAtTimer to duration to wait for this item.
+// If thete are no items in the queue then we set the nextReadyAtTimerCh to a channel which will
+// never fire.
+func (q *delayingType) calculateNextReadyAt(w *waitForPriorityQueue) {
+	q.nextReadyAtTimerCh = make(<-chan time.Time)
+	if w.Len() > 0 {
+		if q.nextReadyAtTimer != nil {
+			q.nextReadyAtTimer.Stop()
```

Comment:
We only enter this section when there are items in the queue

Reply:
yes, but it would be better for the function consistency. For example when we remove item from the queue before its timer is up, we could stop that item timer here even when the queue is empty + set it to nil

Reply:
Ok I see what you mean now, have a commit to move the nextReadyAtTimer outside of the queue length check - thanks!

Reply:
@davidmccormick thx, can you also set the `nextReadyAtTimer` to nil when stopped.  It is not strictly required as per order of operations (after remove there should be an insert), but would look nicer and save us one stop call.

Reply:
ok have done!